### PR TITLE
Model Verifier base

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR brings the base for the `model-verifier` implementation.

The `model-verifier` itself should be placed to `core-java` repository as it is highly coupled with the `server` module.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    private static void processModel(SpineModel model, Project project) {
+        final ProcessingStage validation = ProcessingStages.validate(project);
+        validation.process(model);
+    }
+
+    /**
+     * The action performing the model processing.
+     *
+     * <p>The action is executed only is the passed {@code rawModelPath} is present.
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java]: `if`

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        final Path result = rootDir.resolve(RELATIVE_RAW_MODEL_PATH);
+        return result;
+    }
+
+    private static Action<Task> action(Path path) {
+        return new GeneratorAction(path);
+    }
+
+    /**
+     * Processes the {@link SpineModel} upon the {@linkplain Project Gradle project}.
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java]: I have no idea what you mean here. In particular `Processes` is too generic, `upon` seems to be just incorrect.

Please rewrite.

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * <p>Currently, performs only the model validation. The behavior may change in future.
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ModelGeneratorPlugin.java]: As discussed, let this class be just a `ModelVerifierPlugin`, and this method is then `verifyModel`.

You may be able to create a proper documentation for that.

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStage.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+package io.spine.gradle.compiler.model;
+
+import io.spine.tools.model.SpineModel;
+
+/**
+ * The stage of processing of a {@link SpineModel}.
+ *
+ * @author Dmytro Dashenkov
+ */
+interface ProcessingStage {
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStage.java]: This should be renamed re-documented according to the vision of `just a model verification for now`.

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStages.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            for (String commandHandlingClass : rawModel.getCommandHandlingTypesList()) {
+                final Class<?> cls;
+                try {
+                    log().debug("Trying to load class \'{}\'", commandHandlingClass);
+                    cls = getModelClass(commandHandlingClass);
+                } catch (ClassNotFoundException e) {
+                    throw new IllegalStateException(e);
+                }
+                if (Aggregate.class.isAssignableFrom(cls)) {
+                    @SuppressWarnings("unchecked") final Class<? extends Aggregate> aggregateClass =
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStages.java]: Please have a single `@SuppressWarnings` for the method and a valid explanation.

File: gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStages.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                            (Class<? extends Aggregate>) cls;
+                    model.asAggregateClass(aggregateClass);
+                    log().debug("\'{}\' classified as Aggregate type.");
+                } else if (ProcessManager.class.isAssignableFrom(cls)) {
+                    @SuppressWarnings("unchecked") final Class<? extends ProcessManager> aggregateClass =
+                            (Class<? extends ProcessManager>) cls;
+                    model.asProcessManagerClass(aggregateClass);
+                    log().debug("\'{}\' classified as ProcessManages type.");
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/main/java/io/spine/gradle/compiler/model/ProcessingStages.java]: I am not sure I understand why you need the escaped `{}`.

File: gradle-plugins/model-compiler/src/test/java/io/spine/gradle/compiler/model/ModelGeneratorPluginShould.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                "io/spine/tools/model/DuplicateAggregate.java",
+                "io/spine/tools/model/DuplicateCommandHandler.java")
+                .executeAndFail(GENERATE_MODEL);
+        final BuildTask task = result.task(toPath(GENERATE_MODEL));
+        assertNotNull(task);
+        final TaskOutcome generationResult = task.getOutcome();
+        assertEquals(FAILED, generationResult);
+    }
+
+    @Ignore // TODO:2017-08-25:dmytro.dashenkov: Re-enable when Model is capable of checking the handler methods.
```
Comment (armiol) [inline gradle-plugins/model-compiler/src/test/java/io/spine/gradle/compiler/model/ModelGeneratorPluginShould.java]: Please reference an issues instead.

**Comment by armiol** [comment]
@dmdashenkov I also expect that build succeeds before we merge anything.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 8]
```diff
@@ -158,6 +168,53 @@ public Builder insertAfterTask(TaskName target) {
             }
 
             /**
+             * Specify tasks which will precede the new one.
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: I think it `Inserts`, not `Specify`.

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 9]
```diff
@@ -158,6 +168,53 @@ public Builder insertAfterTask(TaskName target) {
...
+             *
+             * <p>Unlike {@link #insertAfterTask insertAfterTask()}, this method will depend
+             * the new task on <b>every</b> task with such name in the project (i.e. the tasks of
+             * the root project and all the subprojects).
+             *
+             * <p>If a certain project does not have a task with the specified name, no action is
+             * performed for that project.
+             *
+             * <p>This method does not guarantee that the task will be depended on either task.
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: This isn't clear. That's the first time "to depend" is used, maybe it's a root issue.

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 10]
```diff
@@ -158,6 +168,53 @@ public Builder insertAfterTask(TaskName target) {
...
+             * @return the current instance of {@link Builder}
+             */
+            public Builder insertAfterAllTasks(TaskName target) {
+                checkNotNull(target, "tasks before the new one");
+                this.previousTaskOfAllprojects = target;
+                return this;
+            }
+
+            /**
+             * Specify the files or/and directories which are the input data for the new task.
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: `Adds the files and/or directories to the input dataset for the task being built.`

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 11]
```diff
@@ -158,6 +168,53 @@ public Builder insertAfterTask(TaskName target) {
...
+             *
+             * <p>If none of the specified file system elements are present before the task
+             * execution, the task will be marked as {@code NO-SOURCE} and skipped.
+             *
+             * <p>Read more about the Gradle incremental builds
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: This is totally out of the blue. Please don't do that to the potential readers, that looks totally unrelated.

File: gradle-plugins/plugin-base/src/test/java/io/spine/gradle/GradleProject.java [hunk 12]
```diff
@@ -73,24 +76,48 @@ private GradleProject(Builder builder) throws IOException {
...
 
+    private void writeJavaFiles(Iterable<String> fileNames) throws IOException {
+        for (String javaFile : fileNames) {
+            writeJava(javaFile);
+        }
+    }
+
     public BuildResult executeTask(TaskName taskName) {
-        return gradleRunner.withArguments(taskName.getValue())
+        return gradleRunner.withArguments(taskName.getValue(), STACKTRACE_CLI_OPTION, "--debug")
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/test/java/io/spine/gradle/GradleProject.java]: Please introduce the similar constant for `--debug`.

File: gradle-plugins/plugin-base/src/test/java/io/spine/gradle/GradleProject.java [hunk 13]
```diff
@@ -145,11 +173,17 @@ public Builder setProjectFolder(TemporaryFolder folder) {
...
 
         public Builder addProtoFile(String protoFileName) {
-            checkArgument(!isBlank(protoFileName));
+            checkArgument(!isNullOrEmpty(protoFileName));
             protoFileNames.add(protoFileName);
             return this;
         }
 
+        @SuppressWarnings("unused") // Part of API (used in other Spine modules)
+        public Builder addJavaFiles(String... fileNames) {
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/test/java/io/spine/gradle/GradleProject.java]: As long as it was reported "unused" I assume there were no unit tests for this feature.

How do we ever know it works as expected?

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 14]
```diff
@@ -168,25 +225,59 @@ public GradleTask applyNowTo(Project project) {
...
-                final GradleTask result = new GradleTask(name, project);
-                return result;
+                if (previousTaskOfAllprojects != null) {
+                    final Project root = project.getRootProject();
+                    dependTaskOnAllProjects(task, root);
+                }
+            }
+
+            private void dependTaskOnAllProjects(final Task task, Project rootProject) {
+                final String prevTaskName = previousTaskOfAllprojects.getValue();
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: It's weird how the coverage of this class is partially reported. I would completely understand if there were no coverage at all (due to a forked test execution), but to have a partial coverage is a bit out there. Let's discuss.

File: base/src/main/java/io/spine/string/MapStringifier.java [hunk 15]
```diff
@@ -169,7 +168,7 @@ protected String toString(Map<K, V> obj) {
             }
             return resultMap;
         } catch (Throwable e) {
-            throw newIllegalArgumentException("The exception occurred during the conversion", e);
+            throw new IllegalArgumentException("The exception occurred during the conversion", e);
```
Comment (armiol) [inline base/src/main/java/io/spine/string/MapStringifier.java]: Let's discuss why it is changed.

**Comment by armiol** [comment]
@dmdashenkov I would also name the PR properly — according to what it is now.

File: base/src/main/java/io/spine/string/MapStringifier.java (hunk 15 above)
Comment (dmdashenkov) [reply base/src/main/java/io/spine/string/MapStringifier.java]: `newIllegalArgumentException(String, Object)` is tricky, as is supposes that we use it for error message formatting. Instead, here we add the `cause` Throwable to the new one.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java [hunk 16]
```diff
@@ -168,25 +223,59 @@ public GradleTask applyNowTo(Project project) {
...
+                }
+            }
+
+            private void dependTaskOnAllProjects(final Task task, Project rootProject) {
+                final String prevTaskName = previousTaskOfAllProjects.getValue();
+                ProjectHierarchy.applyToAll(rootProject, new Action<Project>() {
+                    @Override
+                    public void execute(Project project) {
+                        final Task existingTask = project.getTasks().findByName(prevTaskName);
+                        if (existingTask != null) {
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: This line is reported to be only partially tested.

File: gradle-plugins/plugin-base/src/test/java/io/spine/gradle/SpinePluginBuilderShould.java [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        final Task task = project.getTasks().findByPath(PRE_CLEAN.getValue());
+        final File singleInput = task.getInputs()
+                                      .getFiles()
+                                      .getFiles()
+                                      .iterator()
+                                      .next();
+        assertEquals(input.getCanonicalFile(), singleInput.getCanonicalFile());
+    }
+
+    private static class TestPlugin extends SpinePlugin {
```
Comment (armiol) [inline gradle-plugins/plugin-base/src/test/java/io/spine/gradle/SpinePluginBuilderShould.java]: With this name this class should at least have a class-level comment on the purpose of creation.

File: gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java (hunk 16 above)
Comment (dmdashenkov) [reply gradle-plugins/plugin-base/src/main/java/io/spine/gradle/SpinePlugin.java]: Covered both cases.
